File: python/src/epostak/resources/firms.py
"""Firms resource -- manage client firms (integrator keys).

Provides :class:`FirmsResource` for listing, inspecting, and assigning firms
when using an integrator API key (``sk_int_*``).  Single-firm keys
(``sk_live_*``) only have access to :meth:`~FirmsResource.list` and
:meth:`~FirmsResource.get`.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, TYPE_CHECKING
from urllib.parse import quote

if TYPE_CHECKING:
    from epostak.types import (
        AssignFirmResponse,
        BatchAssignFirmsResponse,
        FirmDetail,
        FirmSummary,
        InboxListResponse,
        PeppolIdentifierResponse,
    )

from epostak.resources.documents import _BaseResource, _build_query


class FirmsResource(_BaseResource):
    """Manage client firms (integrator keys)."""

    def list(self) -> List[FirmSummary]:
        """List all accessible firms.

        Returns:
            List of firm summary dicts with ``id``, ``name``, ``ico``, ``peppolId``,
            and ``peppolStatus``.

        Example::

            firms = client.firms.list()
            for firm in firms:
                print(firm["name"], firm["peppolStatus"])
        """
        res = self._request("GET", "/firms")
        return res.get("firms", res) if isinstance(res, dict) else res

    def get(self, id: str) -> FirmDetail:
        """Get firm detail by ID.

        Args:
            id: Firm UUID.

        Returns:
            Full firm detail including address, Peppol identifiers, and tax IDs.

        Example::

            firm = client.firms.get("firm-uuid")
            print(firm["name"], firm["peppolIdentifiers"])
        """
        return self._request("GET", f"/firms/{quote(id, safe='')}")

    def documents(
        self,
        id: str,
        offset: int = 0,
        limit: int = 20,
        direction: Optional[str] = None,
    ) -> InboxListResponse:
        """List documents for a firm.

        Args:
            id: Firm UUID.
            offset: Pagination offset (default 0).
            limit: Page size (default 20).
            direction: ``"inbound"`` or ``"outbound"``.

        Returns:
            Paginated response with ``documents``, ``total``, ``limit``, and ``offset``.

        Example::

            docs = client.firms.documents("firm-uuid", limit=50, direction="inbound")
            print(docs["total"], "total documents")
        """
        params = _build_query({"offset": offset, "limit": limit, "direction": direction})
        return self._request("GET", f"/firms/{quote(id, safe='')}/documents", params=params)

    def register_peppol_id(self, id: str, scheme: str, identifier: str) -> PeppolIdentifierResponse:
        """Register a Peppol ID for a firm.

        Args:
            id: Firm UUID.
            scheme: Peppol scheme. For Slovak participants use ``"0245"`` (DIČ) — per Slovak PASR the ``"9950"`` VAT-number form is not supported.
            identifier: Peppol identifier value, e.g. ``"12345678"``.

        Returns:
            Dict with ``peppolId``, ``scheme``, ``identifier``, and ``registeredAt``.

        Example::

            result = client.firms.register_peppol_id("firm-uuid", "0245", "12345678")
            print(result["peppolId"])  # "0245:12345678"
        """
        return self._request(
            "POST",
            f"/firms/{quote(id, safe='')}/peppol-identifiers",
            json={"scheme": scheme, "identifier": identifier},
        )

    def assign(self, ico: str) -> AssignFirmResponse:
        """Link this integrator to a Firm that already completed FS SR signup.

        **Lookup-only** -- this endpoint cannot create new Firms. The target
        Firm must have completed FS SR PFS signup and granted consent to this
        integrator before the link succeeds.

        Args:
            ico: Slovak ICO (8 digits).

        Returns:
            Dict with ``firm`` (linked firm details) and ``status`` (e.g. ``"active"``).

        Raises:
            EPostakError: 404 with ``code="FIRM_NOT_REGISTERED"`` -- no Firm with
                that ICO exists yet. Direct the firm to complete FS SR PFS signup
                before retrying.
            EPostakError: 403 with ``code="CONSENT_REQUIRED"`` -- Firm exists
                but has not granted consent for this integrator.
            EPostakError: 409 with ``code="ALREADY_LINKED"`` -- the integrator
                already has an active link to this Firm.

        Example::

            try:
                result = client.firms.assign("12345678")
                print(result["firm"]["id"], result["status"])
            except EPostakError as err:
                if err.code == "FIRM_NOT_REGISTERED":
                    # ask the firm to complete FS SR PFS signup first
                    ...
        """
        return self._request("POST", "/firms/assign", json={"ico": ico})

    def assign_batch(self, icos: List[str]) -> BatchAssignFirmsResponse:
        """Batch assign firms to this integrator (max 50).

        Args:
            icos: List of Slovak ICOs (8-digit strings).

        Returns:
            Dict with ``results`` -- one entry per ICO with ``status``, ``firm``, and
            optional ``error``/``message`` on failure.

        Example::

            result = client.firms.assign_batch(["12345678", "87654321"])
            for r in result["results"]:
                print(r["ico"], r["status"])
        """
        return self._request("POST", "/firms/assign/batch", json={"icos": icos})

File: ruby/lib/epostak/resources/firms.rb
# frozen_string_literal: true

require "erb"

module EPostak
  module Resources
    # Resource for managing firms (companies) associated with your account.
    # Integrators use this to assign client firms, view their documents,
    # and register Peppol identifiers.
    #
    # @example
    #   firms = client.firms.list
    #   firms["firms"].each { |f| puts "#{f['name']} (#{f['peppolStatus']})" }
    class Firms
      # @param http [EPostak::HttpClient] Internal HTTP client
      def initialize(http)
        @http = http
      end

      # List all firms associated with the current account.
      # For integrator keys, returns all assigned client firms.
      #
      # @return [Hash] Response with "firms" array of firm summaries
      #
      # @example
      #   response = client.firms.list
      #   response["firms"].each { |f| puts f["name"] }
      def list
        @http.request(:get, "/firms")
      end

      # Get detailed information about a specific firm, including tax IDs,
      # address, and all registered Peppol identifiers.
      #
      # @param id [String] Firm UUID
      # @return [Hash] Detailed firm information
      #
      # @example
      #   firm = client.firms.get("firm-uuid")
      #   puts firm["icDph"] # => "SK2020123456"
      def get(id)
        @http.request(:get, "/firms/#{encode(id)}")
      end

      # List documents belonging to a specific firm.
      # Useful for integrators to view a client's document history.
      #
      # @param id [String] Firm UUID
      # @param offset [Integer] Number of documents to skip (default: 0)
      # @param limit [Integer] Maximum documents to return (default: 20)
      # @param direction [String, nil] Filter by "inbound" or "outbound"
      # @return [Hash] Paginated list of the firm's documents
      #
      # @example
      #   response = client.firms.documents("firm-uuid", direction: "inbound", limit: 100)
      #   response["documents"].each { |doc| puts doc["id"] }
      def documents(id, offset: 0, limit: 20, direction: nil)
        query = { offset: offset, limit: limit, direction: direction }
        @http.request(:get, "/firms/#{encode(id)}/documents", query: query)
      end

      # Register a new Peppol identifier for a firm. This enables the firm
      # to send and receive documents on the Peppol network under this identifier.
      #
      # Slovak Peppol ID format: "0245:DIC" (e.g. "0245:1234567890"). Per
      # Slovak PASR, only the 0245 scheme is used — "9950:SK..." is not
      # supported.
      #
      # @param id [String] Firm UUID
      # @param scheme [String] Peppol identifier scheme (e.g. "0245")
      # @param identifier [String] Identifier value within the scheme (e.g. "1234567890")
      # @return [Hash] The registered identifier details
      #
      # @example
      #   result = client.firms.register_peppol_id("firm-uuid", scheme: "0245", identifier: "1234567890")
      #   puts result["peppolId"] # => "0245:1234567890"
      def register_peppol_id(id, scheme:, identifier:)
        @http.request(:post, "/firms/#{encode(id)}/peppol-identifiers", body: {
          scheme: scheme,
          identifier: identifier
        })
      end

      # Assign a firm to the integrator account by its Slovak ICO.
      # Once assigned, you can send/receive documents on behalf of this firm.
      #
      # @param ico [String] Slovak ICO (8-digit business registration number)
      # @return [Hash] The assigned firm details and status
      #
      # @example
      #   result = client.firms.assign(ico: "12345678")
      #   puts result["firm"]["id"] # => Use this UUID for subsequent operations
      def assign(ico:)
        @http.request(:post, "/firms/assign", body: { ico: ico })
      end

      # Assign multiple firms at once by their Slovak ICOs.
      # Each ICO is processed independently -- individual failures don't affect others.
      # Maximum 50 ICOs per request.
      #
      # @param icos [Array<String>] Array of Slovak ICOs
      # @return [Hash] Individual results for each ICO (success or error)
      #
      # @example
      #   result = client.firms.assign_batch(icos: ["12345678", "87654321", "11223344"])
      #   result["results"].each do |r|
      #     if r["error"]
      #       puts "#{r['ico']}: #{r['message']}"
      #     else
      #       puts "#{r['ico']}: #{r['status']}"
      #     end
      #   end
      def assign_batch(icos:)
        @http.request(:post, "/firms/assign/batch", body: { icos: icos })
      end

      private

      def encode(value)
        ERB::Util.url_encode(value)
      end
    end
  end
end
